interfaces/cython/cantera/composite.py

```diff
@@ -516,32 +516,38 @@ def __init__(self, phase, shape=(0,), states=None, extra=None):
         reserved = self.__dir__()
 
         self._extra = {}
-        if isinstance(extra, dict):
-            for name, v in extra.items():
+
+        if isinstance(extra, str):
+            extra = [extra]
+
+        if isinstance(extra, (list, tuple, np.ndarray)) and all(isinstance(name, str) for name in extra):
```

Comment:
I'm not a fan of this approach of trying to exhaustively check the types of an object (referring to the first `isinstance()` check). For instance, I may want to supply the keys or items of a dictionary as an iterable to this function, but:

```python
>>> isinstance({'a': 'b'}.keys(), (list, tuple))
False
```

You might want to do this if you already have a dictionary of data, and you want to reuse just the keys but replace the data. Another example here would be the columns of a Pandas DataFrame, which may be more likely since we have the ability to convert `SolutionArray`s to Pandas DataFrames.

The solution to this problem that I see online is to do:

```python
try:
    iterator = iter(extra)
except TypeError:
    raise ValueError(...) from None
for name in iterator:
    ...
```

Unfortunately, dictionaries would pass this test, but can't be operated on in the same way as a simple sequence like a list (hence the existing separated handling). So here's what I propose:

1. Move the `isinstance(extra, dict)` check to be the first check. I can't think of other objects in Python that behave like dictionaries (i.e., provide mappings) other than perhaps DataFrames. Adding support for DataFrames seems like a sufficiently separate thing that I think it should not be included here, so we can leave the `dict` check as-is (other than my comment on the line below).
2. Have an `else` that does the `try...except...` that I mentioned above. Just watch out that if you do the `all(isinstance(name, str)...)` check on the iterator object that you create from `iter()` that the iterator will be exhausted and you'll have to make a new one, which might be expensive. That said, you could maybe add a separate `assert` like this:
   ```python
   try:
       iterator = iter(extra)
       assert all(isinstance(name, str) for name in extra)
   except TypeError, AssertionError:
       raise ValueError(...) from None
```
In the first line of the `try`, you've already confirmed that the object is, in fact, iterable, so the generator on the assert line should be fine.

Another check that might be good to add somewhere here is if the type is an `np.ndarray`, it is one-dimensional.

Reply:
BTW, the `raise ValueError(...) from None`, the `from None` is necessary to avoid a double traceback, since we don't actually care about the `TypeError` or `AssertionError`, which would otherwise be raised in addition to the `ValueError`.

Reply:
This is certainly insightful. Here is a bit I think about the second part though.

> 2\. Have an `else` that does the `try...except...` that I mentioned above. Just watch out that if you do the `all(isinstance(name, str)...)` check on the iterator object that you create from `iter()` that the iterator will be exhausted and you'll have to make a new one, which might be expensive

we are iterating twice over the entire the iterable. I think instead of first checking and then allocating extra properties we can do that simultaneously and raise an error if that is not a string? something like
```
try:
    iter(extra)
    for name in extra:
        if (isinstance(name, str):
        #allocate extra array
       else:
       #raise error
except TypeError, AssertionError:
    raise ValueError(...) from None

```

Reply:
OK, I like that suggestion! In that case, you can `raise` a `TypeError` in the `else` clause, and you don't need the `AssertionError`.